src/textual/_node_list.py

```diff
@@ -73,8 +90,19 @@ def _insert(self, index: int, widget: Widget) -> None:
         if widget not in self._nodes_set:
             self._nodes.insert(index, widget)
             self._nodes_set.add(widget)
+            widget_id = widget.id
+            if widget_id is not None:
+                self._ensure_unique_id(widget_id)
+                self._nodes_by_id[widget_id] = widget
             self._updates += 1
 
+    def _ensure_unique_id(self, widget_id: str) -> None:
+        if widget_id in self._nodes_by_id:
+            raise DuplicateIds(
+                f"Tried to insert a widget with ID {widget_id!r}, but a widget {self._nodes_by_id[widget_id]!r} "
+                f"already exists with that ID. Widget IDs must be unique."
```

Comment:
Probably worth specifying that widgets should be unique within a given children list. Could you rephrase to cover that?

Reply:
I've reworded it